test/jsonpointer/jsonpointer_weakpointer_test.cc

```diff
@@ -0,0 +1,272 @@
+  EXPECT_TRUE(pointer.at(0).is_property());
+  EXPECT_EQ(pointer.at(0).to_property(), "foo");
+}
+
+TEST(JSONWeakPointer_pointer, pop_back_empty) {
+  sourcemeta::jsontoolkit::WeakPointer pointer;
+  EXPECT_EQ(pointer.size(), 0);
+  EXPECT_THROW(pointer.pop_back(), std::runtime_error);
+}
+
+TEST(JSONWeakPointer_pointer, ordering_less_than) {
+  const sourcemeta::jsontoolkit::WeakPointer pointer_1{std::cref(foo),
+                                                       std::cref(bar)};
+  const sourcemeta::jsontoolkit::WeakPointer pointer_2{std::cref(foo)};
+  const sourcemeta::jsontoolkit::WeakPointer pointer_3{std::cref(baz)};
+  EXPECT_TRUE(pointer_2 < pointer_1);
+  EXPECT_TRUE(pointer_3 < pointer_1);
+  EXPECT_TRUE(pointer_3 < pointer_2);
+}
+
+TEST(JSONWeakPointer_pointer, pop_back_zero_empty) {
+  sourcemeta::jsontoolkit::WeakPointer pointer{};
+  pointer.pop_back(0);
+  EXPECT_EQ(pointer.size(), 0);
+}
+
+TEST(JSONWeakPointer_pointer, pop_back_many_subset) {
+  sourcemeta::jsontoolkit::WeakPointer pointer{std::cref(foo), std::cref(bar),
+                                               std::cref(baz)};
+  pointer.pop_back(2);
+  EXPECT_EQ(pointer.size(), 1);
+  EXPECT_TRUE(pointer.at(0).is_property());
+  EXPECT_EQ(pointer.at(0).to_property(), "foo");
+}
+
+TEST(JSONWeakPointer_pointer, pop_back_many_all) {
+  sourcemeta::jsontoolkit::WeakPointer pointer{std::cref(foo), std::cref(bar),
+                                               std::cref(baz)};
+  pointer.pop_back(3);
+  EXPECT_EQ(pointer.size(), 0);
+}
+
+TEST(JSONWeakPointer_pointer, push_back_pointer_copy) {
+  sourcemeta::jsontoolkit::WeakPointer pointer{std::cref(foo), std::cref(bar)};
+  const sourcemeta::jsontoolkit::WeakPointer other{std::cref(baz),
+                                                   std::cref(qux)};
+  pointer.push_back(other);
+  EXPECT_EQ(pointer.size(), 4);
+  EXPECT_TRUE(pointer.at(0).is_property());
+  EXPECT_EQ(pointer.at(0).to_property(), "foo");
+  EXPECT_TRUE(pointer.at(1).is_property());
+  EXPECT_EQ(pointer.at(1).to_property(), "bar");
+  EXPECT_TRUE(pointer.at(2).is_property());
+  EXPECT_EQ(pointer.at(2).to_property(), "baz");
+  EXPECT_TRUE(pointer.at(3).is_property());
+  EXPECT_EQ(pointer.at(3).to_property(), "qux");
+}
+
+TEST(JSONWeakPointer_pointer, push_back_pointer_move) {
+  sourcemeta::jsontoolkit::WeakPointer pointer{std::cref(foo), std::cref(bar)};
+  sourcemeta::jsontoolkit::WeakPointer other{std::cref(baz), std::cref(qux)};
+
+  pointer.push_back(std::move(other));
+  EXPECT_EQ(pointer.size(), 4);
+  EXPECT_TRUE(pointer.at(0).is_property());
+  EXPECT_EQ(pointer.at(0).to_property(), "foo");
+  EXPECT_TRUE(pointer.at(1).is_property());
+  EXPECT_EQ(pointer.at(1).to_property(), "bar");
+  EXPECT_TRUE(pointer.at(2).is_property());
+  EXPECT_EQ(pointer.at(2).to_property(), "baz");
+  EXPECT_TRUE(pointer.at(3).is_property());
+  EXPECT_EQ(pointer.at(3).to_property(), "qux");
+
+  // TODO: we should assert that other has been moved
```

Comment:
I don't think we are moving correctly here! As there is still stuff in `other`!

Reply:
Its fine if there are still things, but they shouldn't be the same things

Reply:
I'll dig in another PR!